### PR TITLE
Deploying contracts in parts

### PR DESCRIPTION
**Pull request by sinev-valentine**

### CODE REVIEW

File: evm_loader/program/src/entrypoint.rs [hunk 1]
```diff
@@ -275,6 +275,74 @@ fn process_instruction<'a>(
...
+                let trx_len = trx_len.try_into().ok().map(u64::from_le_bytes).unwrap();
+                let (trx, _rest) = rest.split_at(trx_len as usize);
+                (trx.to_vec(), signature.to_vec())
+            };
+            let (nonce, to, data) = get_data(&unsigned_msg);
+            if let Err(e) = verify_tx_signature(&signature, &unsigned_msg) {
+                debug_print!("{}", e);
+                return Err(ProgramError::InvalidInstructionData);
+            }
+            let trx: UnsignedTransaction = rlp::decode(&unsigned_msg).map_err(|_| ProgramError::InvalidInstructionData)?;
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs:305]: Use this or get_data(...). Do not decode transaction twice.

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

File: evm_loader/program/src/entrypoint.rs [hunk 2]
```diff
@@ -275,6 +275,74 @@ fn process_instruction<'a>(
...
+
+            let  accounts = &accounts[1..];
+
+            let (unsigned_msg, signature) = {
+                let data = holder_info.data.borrow();
+                let account_info_data = AccountData::unpack(&data)?;
+                match account_info_data {
+                    AccountData::Empty => (),
+                    _ => return Err(ProgramError::InvalidAccountData),
+                };
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs:292]: Why you want holder_info to be empty?

File: evm_loader/program/src/entrypoint.rs [hunk 3]
```diff
@@ -275,6 +275,74 @@ fn process_instruction<'a>(
...
             Ok(())
         },
+        EvmInstruction::ExecuteTrxFromAccountDataIterative{step_count} =>{
+            debug_print!("Execute iterative transaction from account data");
+            let account_info_iter = &mut accounts.iter();
+            let holder_info = next_account_info(account_info_iter)?;
+            let storage_info = next_account_info(account_info_iter)?;
+            let _program_info = next_account_info(account_info_iter)?;
+            let _program_code = next_account_info(account_info_iter)?;
+            let _caller_info = next_account_info(account_info_iter)?;
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs]: You do not use last three variables. Remove them.

File: evm_loader/program/src/entrypoint.rs [hunk 4]
```diff
@@ -275,6 +275,74 @@ fn process_instruction<'a>(
...
+                debug_print!("Invalid chain id: actual {}, expect {}", trx.chain_id, backend.chain_id());
+                return Err(ProgramError::InvalidInstructionData);
+            }
+
+            let executor_state = ExecutorState::new(ExecutorSubstate::new(), backend);
+            let mut executor = Machine::new(executor_state);
+
+            debug_print!("Executor initialized");
+
+            executor.create_begin(caller_ether, data, u64::max_value());
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs]: Before calling executor you should check that this is a deploy transaction and not a call transaction.

File: evm_loader/program/src/entrypoint.rs [hunk 5]
```diff
@@ -275,6 +275,74 @@ fn process_instruction<'a>(
...
+            executor.execute_n_steps(step_count).unwrap();
+
+            debug_print!("save");
+            executor.save_into(&mut storage);
+            storage.block_accounts(program_id, accounts);
```
Comment (anton-lisanin) [inline evm_loader/program/src/entrypoint.rs]: block_accounts can return an error. Check it.

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

File: evm_loader/program/src/entrypoint.rs (hunk 1 above)
Comment (sinev-valentine) [reply evm_loader/program/src/entrypoint.rs:305]: fixed

File: evm_loader/program/src/entrypoint.rs (hunk 2 above)
Comment (sinev-valentine) [reply evm_loader/program/src/entrypoint.rs:292]: holder account is created by system instruction createAccountWithSeed()

File: evm_loader/program/src/entrypoint.rs (hunk 3 above)
Comment (sinev-valentine) [reply evm_loader/program/src/entrypoint.rs]: unused variables  removed

File: evm_loader/program/src/entrypoint.rs (hunk 4 above)
Comment (sinev-valentine) [reply evm_loader/program/src/entrypoint.rs]: added check trx.to field

File: evm_loader/program/src/entrypoint.rs (hunk 5 above)
Comment (sinev-valentine) [reply evm_loader/program/src/entrypoint.rs]: removed OK() after block_accounts

File: evm_loader/program/src/executor.rs [hunk 6]
```diff
@@ -347,7 +347,36 @@ impl<'config, B: Backend> Machine<'config, B> {
...
+                if self.executor.config.create_increase_nonce {
+                    self.executor.state.inc_nonce(info.address);
+                }
+
+                let mut instance = evm::Runtime::new(
+                    Rc::new(info.init_code),
+                    Rc::new(Vec::new()),
+                    info.context,
+                    &self.executor.config
+                );
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor.rs:375]: Merge `develop` into your branch and remove `Rc` there

File: evm_loader/program/src/executor.rs [hunk 7]
```diff
@@ -347,7 +347,36 @@ impl<'config, B: Backend> Machine<'config, B> {
...
+
+        let scheme = evm::CreateScheme::Legacy {
+            caller: caller,
+        };
+        self.executor.state.enter(u64::max_value(), false);
+
+        match self.executor.create(caller, scheme, U256::zero(),&code, None ){
+            Capture::Exit((reason, address, return_data)) => {
+                debug_print!("create_begin() error ");
+            },
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor.rs:362]: Return an error. And check it in the `entrypoint.rs`.

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

File: evm_loader/program/src/executor.rs (hunk 6 above)
Comment (sinev-valentine) [reply evm_loader/program/src/executor.rs:375]: merge done

File: evm_loader/program/src/executor.rs (hunk 7 above)
Comment (sinev-valentine) [reply evm_loader/program/src/executor.rs:362]: comment was applied

**Review by anton-lisanin — APPROVED** [review]

**Pull request merged**